l10n_br_sale_stock/models/sale_order_line.py

```diff
@@ -14,7 +14,43 @@ def _prepare_procurement_values(self, group_id=False):
         values = self._prepare_br_fiscal_dict()
         values.update(super()._prepare_procurement_values(group_id))
         # Incluir o invoice_state
-        if self.order_id.sale_create_invoice_policy == 'stock_picking':
+        if self.order_id.company_id.sale_create_invoice_policy\
+                == 'stock_picking':
             values['invoice_state'] = '2binvoiced'
 
         return values
+
+    # no trigger product_id.invoice_policy to avoid retroactively changing SO
+    @api.depends(
+        'qty_invoiced', 'qty_delivered', 'product_uom_qty', 'order_id.state')
+    def _get_to_invoice_qty(self):
+        """
+        Compute the quantity to invoice. If the invoice policy is order,
+        the quantity to invoice is calculated from the ordered quantity.
+        Otherwise, the quantity delivered is used.
+        """
+        super()._get_to_invoice_qty()
+
+        for line in self:
+            if line.order_id.state in ['sale', 'done']:
+                if line.product_id.invoice_policy == 'order':
+                    if line.order_id.company_id.sale_create_invoice_policy ==\
+                            'stock_picking' and\
+                            line.product_id.type == 'product':
+                        # O correto seria que ao selecionar
+                        # sale_create_invoice_policy 'stock_picking' os
+                        # produtos tenham o campo invoice_policy definidos para
+                        # 'delivery', porém para evitar que seja criada uma
+                        # Fatura a partir do Pedido de Venda estamos
+                        # alterando isso mesmo para os produtos definidos com
+                        # 'order', já que a Politica de Criação da Fatura no
+                        # caso do Tipo Produto está definida para ser a
+                        # partir do stock.picking .
+                        # TODO: Essa seria a melhor opção ? Por enquanto pelo
+                        #  que vi para ter o mesmo resultado, que é no caso
+                        #  sale_create_invoice_policy 'stock_picking' só ser
+                        #  possível criar a partir do sale.order Faturas das
+                        #  linhas que sejam type service sim, a outra opção
+                        #  seria sobre escrever o metodo action_invoice_create
+                        #  sem ser possível chamar o super.
+                        line.qty_to_invoice = 0
```

Comment:
In this case, the value should be the quantity delivered instead of zero. With the current code, a line with a product which  invoice_police is `'order'` in an order of a company with `'stock_picking'` as sale_create_invoice_policy is never going to be invoiceable.

Reply:
@gabrielcardoso21  está sendo colocado zero porque o objetivo é de não permitir a criação da Fatura do "type" "product" a partir do Pedido de Venda quando a "Politica de Criação de Fatura em Vendas" for o "stock_picking" isso é para forçar a criação apenas através do objeto stock.picking. Olhando o codigo agora esse "if line.product_id.invoice_policy == 'order':" deveria ser "if line.product_id.type == 'product'':" na linha 36 para ficar da forma como descrevi. Podemos debater se esse fluxo deveria ser aplicar apenas quando o produto estiver com  "invoice_policy"  "delivery" e quando for "order" continuar permitindo a criação da Fatura pelo Pedido de Venda ou se realmente é preciso forçar a criação pelo stock.picking em ambos os casos.

Reply:
> Podemos debater se esse fluxo deveria ser aplicar apenas quando o produto estiver com "invoice_policy" "delivery" e quando for "order" continuar permitindo a criação da Fatura pelo Pedido de Venda ou se realmente é preciso forçar a criação pelo stock.picking em ambos os casos.

Eu estava pensando nisso também. Como o Odoo nativamente já faz esta diferenciação seria melhor usar a funcionalidade atual. Mas entendo que provavelmente houve alguma limitação técnica durante o desenvolvimento e que foi necessário forçar pelo menos por enquanto.